### ISSUE TITLE
Package FTBFS on first build for a given Ubuntu series

### ISSUE THREAD

**Issue by mateusrodrigues** [canonical/dotnet-source-build#12]
When a new Ubuntu series comes along, the .NET packages are copied over from the previous series. Currently, the archive has opened for Plucky Puffin and the .NET builds have been copied over from Oracular Oriole. Now, this causes two issues:

1. The builds start to [fail](https://launchpadlibrarian.net/754768946/buildlog_ubuntu-plucky-amd64.dotnet9_9.0.100-9.0.0~rc2-0ubuntu2~ppa1_BUILDING.txt.gz) because of this line in `d/rules:128`:
```
DOTNET_SOURCE_BUILT_ARTIFACTS_TARBALL=($$(find "$(DOTNET_ROOT_DIR)/source-built-artifacts" -maxdepth 1 -name 'Private.SourceBuilt.Artifacts.$(DOTNET_MAJOR).$(DOTNET_MINOR).*.tar.gz')); \
```
On Plucky, it will look for a file called `Private.SourceBuilt.Artifacts.9.0.*.ubuntu.25.04-x64.tar.gz`, which does not exist because the `dotnet-sdk-9.0-source-built-artifacts` binary package was previously build on Oracular Oriole and, therefore, does not include "ubuntu.25.04" in the source-built artifacts tarball name, rather "ubuntu.24.10".

2. Autopkgtests start [failing](https://launchpadlibrarian.net/754768946/buildlog_ubuntu-plucky-amd64.dotnet9_9.0.100-9.0.0~rc2-0ubuntu2~ppa1_BUILDING.txt.gz) because of this line in `d/eng/source_build_artifact_path.py:21`:
```
        globPattern =  (f"{basePath}/Private.SourceBuilt.Artifacts."
                        f"{sdkVersion}*.{runtimeIdentifier}.tar.gz")
```
A first actual build of the package on the new series should fix this issue (such as a no-change rebuild), but problem 1 actually blocks this solution.

I would say that it is safe to remove any references to the RID in the glob patterns, considering that we can in fact get a source-built artifact from a previous series to build the new one, such as is the case for archive openings.

**Comment by dviererbe** [comment]
> I would say that it is safe to remove any references to the RID in the glob patterns, considering that we can in fact get a source-built artifact from a previous series to build the new one, such as is the case for archive openings

Originally we did not match for an exact rid. I added the explicit rid matching to solve issues when multiple `dotnet-sdk-X.0-source-built-artifacts` packages are installed, which is the case if you reuse a chroot for local testing. My intent was also to protect against unwanted build contamination. With the exception of archive opening, the `source-built-artifacts` should match the correct rid and fail if it doesn't. I am not a big fan of simply disabling this protection.

I have an alternative idea. We can look in the changelog file and make the explicit matching optional if it does not match the distribution it is building on. This should enforce rid matching within a cycle and let the build succeed during archive opening.

**Comment by mateusrodrigues** [comment]
> I have an alternative idea. We can look in the changelog file and make the explicit matching optional if it does not match the distribution it is building on.

Yes, I believe this is fine. Just need to make sure the exception is checked on both the rules file and the `d/eng/source_build_artifact_path.py` file to cover both package builds and autopkgtests.

**Comment by dviererbe** [comment]
>  I added the explicit rid matching to solve issues when multiple dotnet-sdk-X.0-source-built-artifacts packages are installed

With the archive opening of Questing (Ubuntu 25.10) and the upcoming March release I thought again about this issue. Re-evaluating my argument; I think this check does not actually provide value. There can be many valid scenarios where the bootstraping¹ SDK does not match the rid of the built artifact. Therefore I keep the strict .NET version matching to prevent using the artifacts of the wrong version, but remove the strict RID matching.

¹ I use this term here in the general sense and do not refer to the initial bootstrapping of .NET in the Ubuntu Archive.
